experimental/ir/ir_symbol.go

```diff
@@ -213,3 +214,168 @@ func (s symtab) sort(c *Context) {
+func (s symtab) lookupBytes(c *Context, fqn []byte) ref[rawSymbol] {
+	id, ok := c.session.intern.QueryBytes(fqn)
+	if !ok {
+		return ref[rawSymbol]{}
+	}
+	idx, ok := slicesx.BinarySearchKey(s, id, func(r ref[rawSymbol]) intern.ID {
+		return wrapSymbol(c, r).InternedFullName()
+	})
+	if !ok {
+		return ref[rawSymbol]{}
+	}
+
+	return s[idx]
+}
+
+// resolve attempts to resolve the relative path name within the given scope
+// (which should itself be a possibly-empty relative path).
+//
+// Returns zero if the symbol is not found. If the symbol is not found due to
+// Protobuf's weird double-lookup semantics around nested identifiers, this
+// function will try to find the name as if this language bug did not exist, and
+// will report the name it had expected to find.
+//
+// If skipIfNot is nil, the symbol's kind will not be checked to determine if
+// we should continue climbing scopes.
+func (s symtab) resolve(
+	c *Context,
+	scope, name FullName,
+	skipIfNot func(SymbolKind) bool,
+) (found ref[rawSymbol], expected FullName) {
+	// This function implements the name resolution algorithm specified at
+	// https://protobuf.com/docs/language-spec#reference-resolution.
+
+	// Symbol resolution is not quite as simple as trying p + name for all
+	// ancestors of scope. Consider the following files:
+	//
+	// 	// a.proto
+	// 	package foo.bar;
+	// 	message M {}
+	//
+	// 	// b.proto
+	//  package foo;
+	//  import "a.proto";
+	//  message M {}
+	//
+	//  // c.proto
+	//  package foo.bar.baz;
+	//  import "b.proto";
+	//  message N {
+	//    M m = 1;
+	//  }
+	//
+	// The candidates, in order, are:
+	// - foo.bar.baz.M; does not exist.
+	// - foo.bar.M; not visible.
+	// - foo.M; correct answer.
+	// - M; not tried.
+	//
+	// If we do not keep going after encountering symbols that are not visible
+	// to us, we will reject valid code.
+
+	// A similar situation happens here:
+	//
+	// package foo;
+	// message M {
+	//   message N {}
+	//   message P {
+	//     enum X { N = 1; }
+	//     N n = 1;
+	//   }
+	// }
+	//
+	// If we look up N, the candidates are foo.M.P.N, foo.M.N, foo.N, and N.
+	// We will find foo.M.P.N, which is not a message or enum type, so we must
+	// skip it to find the correct name, foo.M.N. This is what the accept
+	// predicate is for.
+
+	// Finally, consider the following situation, which involves partial
+	// names.
+	//
+	// package foo;
+	// message M {
+	//   message N {}
+	//   message M {
+	//     M.N n = 1;
+	//   }
+	// }
+	//
+	// The candidates are foo.M.M.N, foo.M.N, M.N. However, protoc rejects this,
+	// because it actually searches for M first, and then appends the rest of
+	// the path and searches for that, in two phases.
+	//
+	// It is not clear why protoc does this, but it does mean we need to be
+	// careful in how we resolve partial names.
+
+	scopeSearch := !name.IsIdent()
+	first := name.First()
+	candidate := []byte(scope.Append(first))
```

Comment:
Why create a separate byte slice for this? Maybe `symtab.lookupBytes` should instead be `symtab.lookupFullName`? Or maybe this could use `unsafe` to get a backing slice? The only mutation is an `append` call below, but that is safe if the slice that is created via unsafe has its length==capacity, since it will always allocate a new backing array for that case.

Reply:
This is mutated by both delete-in-the-middle and append, so we are actually forced to make this defensive copy. Trust me when I say that I tried to get rid of it.

That said, I can avoid the allocation using sync.Pool. Hooray!